doc/developers/cython.rst

```diff
@@ -0,0 +1,143 @@
+         cythonX --annotate source.pyx
+
+* Using the ``--annotate`` option with this flag allows generating a HTML report of code annotation.
+    This report indicates interactions with the CPython interpreter on a line-by-line basis.
+    Interactions with the CPython interpreter must be avoided as much as possible in
+    the computationally intensive sections of the algorithms.
+    For more information, please refer to `this section of Cython's tutorial <https://cython.readthedocs.io/en/latest/src/tutorial/cython_tutorial.html#primes>`_
+
+    .. code-block::
+
+         # This generates a HTML report (`source.html`) for `source.c`.
+         cythonX --annotate source.pyx
+
+Tips for performance
+^^^^^^^^^^^^^^^^^^^^
+
+* Understand the GIL in context for CPython (which problems it solves, what are its limitations)
+  and get a good understanding of when Cython will be mapped to C code free of interactions with
+  CPython, when it will not, and when it cannot (e.g. presence of interactions with Python
+  objects, which include functions). In this regard, `PEP073 <https://peps.python.org/pep-0703/>`_
+  provides a good overview and context and pathways for removal.
+
+* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.
+
+* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryviews are lightweight.
+
+* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and
+  we better not use it, even if handling fewer dimensions in some context would be preferable.
+
+* Decorate final classes or methods with ``@final`` (this allows removing virtual tables when needed)
+
+* Inline methods and function when it makes sense
+
+* Make sure your Cython compilation units `use NumPy recent C API <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/setup.py#L64-L70>`_.
+
+* In doubt, read the generated C or C++ code if you can: "The fewer C instructions and indirections
+  for a line of Cython code, the better" is a good rule of thumb.
+
+* ``nogil`` declarations are just hints: when declaring the ``cdef`` functions
+  as nogil, it means that they can be called without holding the GIL, but it does not release
+  the GIL when entering them. You have to do that yourself either by passing ``nogil=True`` to
+  ``cython.parallel.prange`` explicitly, or by using an explicit context manager:
+
+.. code-block:: cython
+
+   cdef inline void my_func(self) nogil:
+
+        # Some logic interacting with CPython, e.g. allocating arrays via NumPy.
+
+        with nogil:
+            # The code here is run as is it were written in C.
+
+        return 0
+
+This item is based on `this comment from Stéfan's Benhel <https://github.com/cython/cython/issues/2798#issuecomment-459971828>`_
```

Comment:
Maybe this is the cause of the bad rendering. It should probably be indented

Reply:
Fixed by 7ffdae28775a9a23074c9dc04bd365b55877efe5.

Reply:
it looks great now